### DISCUSSION TITLE
ZkSync Account Abstraction: `forge coverage --via-ir` fails with Stack too deep

### DISCUSSION THREAD

**Discussion by 0x4rm4nd** [Cyfrin/foundry-full-course-cu#1966] [Q&A]
Hi!

At the end of Lesson 15: Foundry Account Abstraction, I'm looking to complete the test suite. However, I encounter a problem when using `forge coverage --via-ir` to compile my Solidity project. The following error occurs:
```bash
minimal-account-abstraction % forge coverage --via-ir
[⠒] Compiling...
[⠰] Compiling 88 files with Solc 0.8.24
[⠑] Solc 0.8.24 finished in 8.39s
Error: 
Compiler run failed:
Error: Compiler error (/solidity/libyul/backends/evm/AsmCodeGen.cpp:67):Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables. When compiling inline assembly: Variable value0 is 2 slot(s) too deep inside the stack. Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables.
CompilerError: Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables. When compiling inline assembly: Variable value0 is 2 slot(s) too deep inside the stack. Stack too deep. Try compiling with `--via-ir` (cli) or the equivalent `viaIR: true` (standard JSON) while enabling the optimizer. Otherwise, try removing local variables.
```
---
### Command executed:
`forge coverage --via-ir`

### Environment
Solidity Version: 0.8.24
Foundry Version: forge 0.0.2 (from `foundry-zksync`)
OS: MacOS 14.2.1

### Context
I also have this problem with the final code of the `minimal-account-abstraction` repository. I run the forge coverage command and get the error mentioned above.

### What I've tried
- Enable optimization: I've tried using `--via-ir --optimize` but the error persists.
- Reduce local variables: I tried refactoring some functions to reduce the number of state variables, but this didn't solve the problem.

---
I read that the error seems to be related to assembler, but I don't see where I can rework the code and modify the stack.
Do you have any suggestions on how to solve this problem or additional leads to explore?

Thanks in advance!

**Comment by EngrPips** [comment]
I don't know how to fix this. Let's look forward to @PatrickAlphaC suggestions when they are available.

**Reply by 0x4rm4nd** [reply]
No worries, thank you for having taken the time to look into this issue 😄

**Reply by EngrPips** [reply]
You are welcome, Friend.

**Comment by SiewLin-AIDeaBuddy** [comment]
I have the same error mesage as well. Trying `forge coverage --ir-minimum` also not working.

It seems to be an on-going issue currently : https://github.com/foundry-rs/foundry/issues/3357

**Reply by EngrPips** [reply]
Interesting. Thanks for the heads-up.

**Comment by PatrickAlphaC** [comment]
@0x5chn0uf can you please follow up here?

https://github.com/matter-labs/foundry-zksync/issues/475

I've added your issue there.

**Comment by Cryptoineer** [comment]
To fix this add ``via-ir = true`` to your foundry.toml file

**Reply by EngrPips** [reply]
indeed
